pkg/mount/magic.go

```diff
@@ -0,0 +1,206 @@
+// Performance is not really an issue: it is a short list, and there are simply
+// not enough block devices/file systems for it to really matter.
+// The ordering for the identical magic number file systems matters: ext4 is more
+// desirable than ext2, so, we want to find ext4 first.
+// The order should NOT BE ALPHABETIC, therefore; it should be ordered with known systems
+// first, and, to break ties, with the most desirable of those systems first.
+var magics = []magic{
+	// From the filesystems magic:
+	// 0x438   leshort         0xEF53          Linux
+	{magic: EXT4, name: "ext4", off: 0x438},
+	{magic: EXT3, name: "ext3", off: 0x438},
+	{magic: EXT2, name: "ext2", off: 0x438},
+	// We will always mount vfat; it's backward compatible (we think?)
+	{magic: MSDOS, name: "vfat", off: 0},
+	{magic: SQUASHFS, name: "squashfs", flags: MS_RDONLY, off: 0},
+	{magic: ISOFS, name: "iso9660", flags: MS_RDONLY, off: 32768},
+	{magic: VFAT, name: "vfat", off: 0},
+	{magic: XFS, name: "xfs", off: 0},
+}
+
+var unknownmagics = []magic{
+	//
+	// here there be dragons.
+	//
+	{magic: V9FS, name: "9p", off: -1},
+	{magic: ADFS, name: "adfs", off: -1},
+	{magic: AFFS, name: "affs", off: -1},
+	{magic: BTRFS, name: "btrfs", off: -1},
+	{magic: SMB, name: "cifs", off: -1},
+	{magic: SMB, name: "smb3", off: -1},
+	{magic: CODA, name: "coda", off: -1},
+	{magic: DEVPTS, name: "devpts", off: -1},
+	{magic: ECRYPTFS, name: "ecryptfs", off: -1},
+	{magic: EFIVARFS, name: "efivarfs", off: -1},
+	{magic: EFS, name: "efs", off: -1},
+	{magic: F2FS, name: "f2fs", off: -1},
+	{magic: FUSE, name: "fuse", off: -1},
+	// ?? {magic: GFS2, name: "gfs2", off: -1},
+	// who care ... {magic: HFSPLUS_VOLHEAD_SIG, name: "hfsplus", off: -1},
+	{magic: HOSTFS, name: "hostfs", off: -1},
+	{magic: HPFS, name: "hpfs", off: -1},
+	{magic: HUGETLBFS, name: "hugetlbfs", off: -1},
+	{magic: JFFS2, name: "jffs2", off: -1},
+	{magic: JFS, name: "jfs", off: -1},
+	{magic: NFS, name: "nfs", off: -1},
+	{magic: NTFS, name: "ntfs", off: -1},
+	{magic: OPENPROM, name: "openpromfs", off: -1},
+	{magic: OVERLAYFS, name: "overlay", off: -1},
+	{magic: PIPEFS, name: "pipefs", off: -1},
+	{magic: PROC, name: "proc", flags: MS_RDONLY, off: -1},
+	{magic: PSTOREFS, name: "pstore", off: -1},
+	{magic: QNX4, name: "qnx4", off: -1},
+	{magic: QNX6, name: "qnx6", off: -1},
+	{magic: RAMFS, name: "ramfs", off: -1},
+	{magic: ROMFS, name: "romfs", flags: MS_RDONLY, off: -1},
+	{magic: UBIFS, name: "ubifs", flags: MS_RDONLY, off: -1},
+	{magic: UDF, name: "udf", off: -1},
+	{magic: ZONEFS, name: "zonefs", off: -1},
+}
+
+// FindMagics finds all the magics matching a magic number.
+func FindMagics(b []byte) []magic {
+	var matches = []magic{}
+	for _, v := range magics {
+		if bytes.Equal(v.magic, b[v.off:v.off+len(v.magic)]) {
+			matches = append(matches, v)
+		}
+	}
+	return matches
+}
+
+// FSFromBlock determines the file system type of a block device.
```

Comment:
Can you add at lease one unit test for this function? Perhaps with a small squashfs. It might need root.

Reply:
I wondered about that but istm it is tested by the caller fairly completely? or not.

Reply:
oh wait there is a unit test for this function ...